terraform/autoneg/main.tf

```diff
@@ -0,0 +1,335 @@
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "clusterrolebinding_autoneg_manager_rolebinding" {
+  metadata {
+    name = "autoneg-manager-rolebinding"
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind      = "ClusterRole"
+    name      = kubernetes_cluster_role.clusterrole_autoneg_manager_role.metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    name      = kubernetes_service_account.service_account.metadata[0].name
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "clusterrolebinding_autoneg_proxy_rolebinding" {
+  metadata {
+    name = "autoneg-proxy-rolebinding"
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind      = "ClusterRole"
+    name      = kubernetes_cluster_role.clusterrole_autoneg_proxy_role.metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    name      = kubernetes_service_account.service_account.metadata[0].name
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+  }
+}
+
+resource "kubernetes_service" "service_autoneg_controller_manager_metrics_service" {
+  metadata {
+    annotations = {
+      "prometheus.io/port"   = "8443"
+      "prometheus.io/scheme" = "https"
+      "prometheus.io/scrape" = "true"
+      "cloud.google.com"     = "'{\"ingress\":true}'"
+    }
+    labels = {
+      "app"           = "autoneg"
+      "control-plane" = "controller-manager"
+    }
+    name      = "autoneg-controller-manager-metrics-service"
+    namespace = "autoneg-system"
+  }
+  spec {
+    port {
+      name        = "https"
+      port        = 8443
+      target_port = "https"
+      protocol    = "TCP"
+    }
+    selector = {
+      app           = "autoneg"
+      control-plane = "controller-manager"
+    }
+  }
+  depends_on = [kubernetes_namespace.namespace_autoneg_system]
+}
+
+resource "kubernetes_deployment" "deployment_autoneg_controller_manager" {
+  metadata {
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+    name      = "autoneg-controller-manager"
+    labels = {
+      "app"           = "autoneg"
+      "control-plane" = "controller-manager"
+    }
+    annotations = {}
+  }
+
+  spec {
+    replicas = 1
+    selector {
+      match_labels = {
+        app           = "autoneg"
+        control-plane = "controller-manager"
+      }
+    }
+
+
+    template {
+      metadata {
+        labels = {
+          app           = "autoneg"
+          control-plane = "controller-manager"
+        }
+        annotations = {}
+      }
+
+      spec {
+        service_account_name             = kubernetes_service_account.service_account.metadata[0].name
+        termination_grace_period_seconds = 10
+
```

Comment:
Explicitly adding `automount_service_account_token  = true` would probably be a good addition here.

Terraform provider changed the default for this from `false` to `true` in a recent version (https://github.com/hashicorp/terraform-provider-kubernetes/pull/1054) so older versions of terraform will require this to be explicitly specified. Plus, if the default has changed once, it may change again... 
The `kubectl` default is `true`, though the more secure default would be `false` since many (most?) workloads don't need API server access and the mounted token becomes a security vulnerability on a compromised workload. `auto-neg` just happens to be one of those workloads that _does_ need the token mounted. `kubectl` isn't changing the default for now since it would be a breaking changed for those workloads that need the token, but aren't explicitly declaring that in their manifests.

Reply:
[terraform-provider-kubernetes v1.13.3](https://github.com/hashicorp/terraform-provider-kubernetes/releases/tag/v1.13.3) and earlier will require setting `automount_service_account_token = true`

[terraform-provider-kubernetes v2.0.0](https://github.com/hashicorp/terraform-provider-kubernetes/releases/tag/v2.0.0) forward, this is currently optional but would be good to be explicit.

Reply:
Good call, I will add this.